### PR TITLE
Turning on flag for FP Reduction Reassociation.

### PR DESCRIPTION
**Pull request by vmurali**

### CODE REVIEW

**Comment by hanhanW** [comment]
I thought we don't want to enable the flag by default?

If we want to track the performance with these experimental flags on, we can add it to our benchmarking set. E.g., https://github.com/iree-org/iree/blob/main/benchmarks/TFLite/linux-x86_64.cmake#L171-L202

Here are the benchmark configuration files related to RISC-V, x86, and mobile CPU:

RISC-V: https://github.com/iree-org/iree/blob/main/benchmarks/TFLite/linux-riscv.cmake
X86: https://github.com/iree-org/iree/blob/main/benchmarks/TFLite/linux-x86_64.cmake and https://github.com/iree-org/iree/blob/main/benchmarks/TF/linux-x86_64.cmake
ARM: https://github.com/iree-org/iree/blob/main/benchmarks/TFLite/android-arm64-v8a.cmake

**Comment by vmurali** [comment]
> I thought we don't want to enable the flag by default?
> 
> If we want to track the performance with these experimental flags on, we can add it to our benchmarking set. E.g., https://github.com/iree-org/iree/blob/main/benchmarks/TFLite/linux-x86_64.cmake#L171-L202
> 
> Here are the benchmark configuration files related to RISC-V, x86, and mobile CPU:
> 
> RISC-V: https://github.com/iree-org/iree/blob/main/benchmarks/TFLite/linux-riscv.cmake X86: https://github.com/iree-org/iree/blob/main/benchmarks/TFLite/linux-x86_64.cmake and https://github.com/iree-org/iree/blob/main/benchmarks/TF/linux-x86_64.cmake ARM: https://github.com/iree-org/iree/blob/main/benchmarks/TFLite/android-arm64-v8a.cmake

Had an offline discussion with @MaheshRavishankar . Let me see if I can turn it on fr benchmarking alone first.

File: benchmarks/TF/linux-x86_64.cmake [hunk 1]
```diff
@@ -20,6 +20,7 @@ set(LINUX_X86_64_CASCADELAKE_CPU_COMPILATION_FLAGS
   "--iree-input-type=mhlo"
   "--iree-llvm-target-cpu=cascadelake"
   "--iree-llvm-target-triple=x86_64-unknown-linux-gnu"
+  "--iree-llvmcpu-reassociate-fp-reductions=true"
```
Comment (hanhanW) [inline benchmarks/TF/linux-x86_64.cmake:23]: Sorry that I did not describe it clearly. There are two categories of benchmark suites. One is default-flags and the other is experimental-flags. We only set target related flags to `default-flags`. Setting the flag here would propagate toe all the benchmark suites.

Can we add `--iree-llvmcpu-reassociate-fp-reductions=true` to experimental-flags benchmark suites only?

**Comment by MaheshRavishankar** [comment]
Let's evaluate what the performance gain is. If we have the flag we can turn it off when we have numerical issues.

File: benchmarks/TF/linux-x86_64.cmake (hunk 1 above)
Comment (MaheshRavishankar) [reply benchmarks/TF/linux-x86_64.cmake:23]: True, but the experimental path on benchmarks is used for MMT4D performance tracking. Totally understand numerical issues are something we need to have a handle on. But this is probably something we should have a flag, but we just turn it on by default cause that is kind of what people expect. If we see numerical issues (or are generating golden values) we could turn this flag off.

**Comment by MaheshRavishankar** [comment]
The numbers don't seem to justify turning it on my default. I would expect this to help though. Is there an idea of what is hampering perf (I am assuming split reduction also is enabled by the same flag)

**Comment by vmurali** [comment]
Perhaps analyse this with tracy first @MaheshRavishankar ?

**Comment by dcaballe** [comment]
What is the current state of unrolling when split reduction is used? Are we not doing any unrolling or vector unrolling or did we finally get scf unrolling to work?

**Comment by hanhanW** [comment]
What models do we expect to see perf improvements? Maybe we can grab some traces for those models and see what's happening.

**Comment by MaheshRavishankar** [comment]
> SplitReduction is enabled by default, yes. Perhaps analyse this with tracy first @MaheshRavishankar ?

SplitReduction also does associative re-ordering. Are they different flags? If we are already enabling split reduction by default (and this is not causing regressions) can we instead use the same flag to turn both on and both off?

**Comment by hanhanW** [comment]
> > SplitReduction is enabled by default, yes. Perhaps analyse this with tracy first @MaheshRavishankar ?
> 
> SplitReduction also does associative re-ordering. Are they different flags? If we are already enabling split reduction by default (and this is not causing regressions) can we instead use the same flag to turn both on and both off?

We enable split reduction by default for integers, not fp. So it is not causing numerical regressions.

We use the same flag to turn both on and both off.

**Comment by MaheshRavishankar** [comment]
> > > SplitReduction is enabled by default, yes. Perhaps analyse this with tracy first @MaheshRavishankar ?
> > 
> > 
> > SplitReduction also does associative re-ordering. Are they different flags? If we are already enabling split reduction by default (and this is not causing regressions) can we instead use the same flag to turn both on and both off?
> 
> We enable split reduction by default for integers, not fp. So it is not causing numerical regressions.
> 
> We use the same flag to turn both on and both off.

So if thats the case I would expect split reduction + associative reordering to help more.... I think using tracy/vtune might give more insight.

**Comment by vmurali** [comment]
> > > > SplitReduction is enabled by default, yes. Perhaps analyse this with tracy first @MaheshRavishankar ?
> > > 
> > > 
> > > SplitReduction also does associative re-ordering. Are they different flags? If we are already enabling split reduction by default (and this is not causing regressions) can we instead use the same flag to turn both on and both off?
> > 
> > 
> > We enable split reduction by default for integers, not fp. So it is not causing numerical regressions.
> > We use the same flag to turn both on and both off.
> 
> So if thats the case I would expect split reduction + associative reordering to help more.... I think using tracy/vtune might give more insight.

One issue is that SplitReduction+AssociativeReordering is already turned on for i32 and i64. This turns it on only for floats.

**Comment by MaheshRavishankar** [comment]
Yeah makes sense. But for float too it should help. I don't think it is controversial to turn this on for integer types. I think LLVM intrinsics don't even have a way to avoid reassociation for integer types.

**Comment by hanhanW** [comment]
> Yeah makes sense. But for float too it should help. I don't think it is controversial to turn this on for integer types. I think LLVM intrinsics don't even have a way to avoid reassociation for integer types.

The improvements from integer types could be better memory access. I expect to see similar improvements on FP side. Especially reductions are mostly on FP for vision models, if I remember correctly.

**Comment by vmurali** [comment]
> Yeah makes sense. But for float too it should help. I don't think it is controversial to turn this on for integer types. I think LLVM intrinsics don't even have a way to avoid reassociation for integer types.

I meant to say that that could be a reason for no speedup

**Comment by dcaballe** [comment]
There are a bunch of x86 benchmarks with 5-7% which seems reasonable to me taking into account that his is only fp standalone reductions. Int reductions should be enabled by default and reductions happening as part of matmuls and other named ops won't be targeted. Also note that we are not comparing apples to apples. Unrolling is disabled when split reduction is used. I limited unrolling for reductions by 8 on RISC-V which means that we should be unrolling by 32 or more on x86. So we are comparing (default reduction strategy + aggressive unrolling) vs (split reduction + no unrolling). Honestly, I'm actually surprised to see improvements with unrolling disabled :). MobileBERT should be an model where this should be kicking in.

Perhaps we should wait until the unrolling issue is fixed to enable this by default for benchmarks? WDYT?

**Comment by MaheshRavishankar** [comment]
Ok. We can wait to evaluate more to enable by default. I just tune out 5-7% as noise, so I might have missed the improvements.

**Comment by dcaballe** [comment]
I also see a 7% improvement on RISC-V for both MobileBERT-float and quant. Reduction was the hottest dispatch with over 7% execution time and goes all the way to 0.49% :). The code is not idea, because it's not unrolled but...
```
  7.49 │       add          s4, s2, s3                                                                                                                                                                ▒
 16.00 │       vle32.v      v11, (s4)                                                                                                                                                                 ▒
  8.08 │       addi         s3, s3, 64                                                                                                                                                                ▒
 61.02 │       vfadd.vv     v10, v10, v11                                                                                                                                                             ◆
  1.31 │     → bne          s3, a2, 0x4fe80 <main_dispatch_33_generic_4x384x384+0xd8>                                                                                                                 ▒
  0.35 │       add          s3, t6, s1                                                                                                                                                                ▒
  0.35 │       slli         s3, s3, 2                                                                                                                                                                 ▒
  0.37 │       add          s3, a1, s3                                                                                                                                                                ▒
       │       vsetivli     zero, 1, e32, m1, ta, ma                                                                                                                                                  ▒
  0.23 │       vlse32.v     v11, (s3), zero                                                                                                                                                           ▒
  0.73 │       vsetivli     zero, 16, e32, m1, ta, ma                                                                                                                                                 ▒
  0.39 │       vfredosum.vs v10, v10, v11                                                                                                                                                             ▒
  0.55 │       vsetivli     zero, 1, e32, m1, ta, ma                                                                                                                                                  ▒
  0.40 │       vse32.v      v10, (s3)
```

However, I still see an ordered horizontal reduction. That should be unordered if the reassoc information is propagated properly. I'll reopen the issue so that we take a look.

**Review by GMNGeoffrey — CHANGES_REQUESTED** [review]
As @pzread says, default flags should not include this. These configs are intended to track the default IREE experience. They only contain flags at all that are strictly necessary for specifying the frontend and backend to use. This was the compromise we made to prevent what was becoming a problem of hand-tuning flags for benchmarks and then having unreproducible performance results. To avoid having too many different configurations (partially a concern due to limited devices to execute on) and deal with the fact that the current benchmark framework indexes benchmarks by these tags, we came up with the "default" vs "experimental" split to allow devs to track performance on something they cared about that couldn't yet be turned on by default. Only having one such config was a known issue and concession to practicalities mentioned above. I'll try to dig up the discussion.

If this should be on by default, that should happen in the code (flip the flag default). If it's architecture dependent, then we should again capture that in the code. Users should not need to know a complex and varying incantation to invoke the compiler.

**Comment by dcaballe** [comment]
> As @pzread says, default flags should not include this. These configs are intended to track the default IREE experience. They only contain flags at all that are strictly necessary for specifying the frontend and backend to use. This was the compromise we made to prevent what was becoming a problem of hand-tuning flags for benchmarks and then having unreproducible performance results. To avoid having too many different configurations (partially a concern due to limited devices to execute on) and deal with the fact that the current benchmark framework indexes benchmarks by these tags, we came up with the "default" vs "experimental" split to allow devs to track performance on something they cared about that couldn't yet be turned on by default. Only having one such config was a known issue and concession to practicalities mentioned above. I'll try to dig up the discussion.
> 
> If this should be on by default, that should happen in the code (flip the flag default). If it's architecture dependent, then we should again capture that in the code. Users should not need to know a complex and varying incantation to invoke the compiler.

That sounds reasonable to me. Once the implementation works to the expected level, the decision about turning the flag on by default would be a matter of whether we want IREE to provide accurate but slower vs less accurate but faster results. I think this should be evaluated by looking at the full context. Our compiler optimization level is pretty aggressive (`-O2`/`-O3`) and I think we are not using high-precision approximations for most of the math primitives. Perhaps, we should think about defining our own set of optimization levels to provide a consistent experience to users with different goals in this regard.

@benvanik

**Comment by hanhanW** [comment]
The android side could be noise. (They are very noisy.)

I think it worth to grab some traces for x86 side. Given that @dcaballe is seeing improvements in mobile bert, we can take a look at the model on x86 side. We should also take a look at MobileNetV2 because the regression might be real.

**Comment by vmurali** [comment]
> However, I still see an ordered horizontal reduction. That should be unordered if the reassoc information is propagated properly. I'll reopen the issue so that we take a look.

Yes, `reassoc` attribute is set to true in `llvm.vector.reduce.fadd`. Sorry this message got lost so didn't reply earlier.

**Comment by benvanik** [comment]
Closing stale PR.

**Pull request closed without merging**